# План вивчення бібліотеки NumPy

1. Основи NumPy:
    * Встановлення та імпортування NumPy.
    * Основи масивів: створення масивів, індексація, зрізи, форми, типи даних.
    * Операції з масивами: арифметичні операції, логічні операції.
   
  
2. Продвинуті Операції з Масивами:
    * Агрегація даних: sum, min, max, mean.
    * Трансформації масивів: reshape, transpose.
    * Об'єднання та розділення масивів.
    
    
3. Операції лінійної алгебри:
    * Множення матриць.
    * Визначники, ранги, сліди матриць.
    * Рішення лінійних систем рівнянь.


4. Обробка Даних:
    * Використання масивів для обробки даних.
    * Функції для статистичного аналізу.


5. Застосування NumPy в реальних задачах:
    * Аналіз даних.
    * Машинне навчання з використанням NumPy.
    * Використання разом з іншими бібліотеками, як-от Pandas, Matplotlib.
  
  
6. Додаткові Матеріали та Ресурси:
    * Офіційна документація NumPy.
    * Онлайн курси та навчальні посібники.
    * Вивчення джерельних кодів бібліотеки.
    
    
7. Практичні завдання та проекти:
    * Реалізація простих проектів для закріплення знань.
    * Вирішення задач з обробки даних та лінійної алгебри.

# Основи NumPy

NumPy, що розшифровується як "Numerical Python", є однією з основних бібліотек для наукових обчислень в Python. Вона надає підтримку великих, багатовимірних масивів і матриць, разом із великою колекцією високорівневих математичних функцій для ефективної роботи з цими даними.

Основні Особливості NumPy:
* Ефективність: NumPy використовує оптимізовані C API для обчислень, що робить його набагато швидшим, ніж вбудовані Python списки для великих об'ємів даних.

* Масиви: Центральним елементом NumPy є n-вимірні масиви (ndarray), які дозволяють виконувати векторизовані операції (операції, виконувані над кожним елементом масиву одночасно), значно підвищуючи ефективність.

* Типи Даних: NumPy підтримує більший діапазон числових типів, ніж Python, що є критично важливим для наукових обчислень.

* Широкий Спектр Математичних Функцій: NumPy містить функції для роботи з основними математичними операціями, статистикою, алгеброю, трансформаціями Фур'є, оптимізаціями та іншими.

* Інтеграція з Іншими Бібліотеками: NumPy інтегрується з багатьма іншими науковими та інженерними бібліотеками Python, такими як SciPy, Pandas, Matplotlib, scikit-learn, scikit-image тощо.

Приклади Використання NumPy:
* Наукові Обчислення: Розрахунок складних математичних формул, включаючи лінійну алгебру, статистику, оптимізацію.

* Обробка Даних / Аналіз Даних: Перетворення та маніпуляції з великими датасетами, включаючи фільтрацію, сортування, групування.

* Машинне Навчання: Використання для зберігання та обробки даних, на яких базуються моделі машинного навчання.

* Обробка Зображень та Сигналів: Застосування для аналізу та модифікації зображень і сигналів.


### Встановлення та Імпортування NumPy
Встановіть NumPy через pip:

```bash
pip install numpy
```

Імпортуйте бібліотеку:

In [2]:
import numpy as np

### Створення масивів:

Одновимірні масиви: 

In [4]:
arr = np.array([1, 2, 3])
print(arr)

[1 2 3]


Двовимірні масиви: 

In [5]:
arr = np.array([[1, 2, 3], [4, 5, 6]])
print(arr)

[[1 2 3]
 [4 5 6]]


Спеціальні масиви: 

In [7]:
arr = np.zeros((2,2)) # створити матрицю і заповнити її нулями
print(arr)

[[0. 0.]
 [0. 0.]]


In [6]:
arr = np.ones((3,3)) # створити матрицю і заповнити її одиницями
print(arr)

[[1. 1. 1.]
 [1. 1. 1.]
 [1. 1. 1.]]


In [8]:
arr = np.eye(3) # одинична матриця
print(arr)

[[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]


In [11]:
arr = np.arange(start=1, stop=9, step=2)  # анафлог range в python
print(arr)

[1 3 5 7]


### Індексація та Зрізи:

Отримання елементу: 
    

In [13]:
arr = np.array([1, 2, 3])
print("Елемент під індексом 2", arr[2])
print("Останній елемент", arr[-1])

Елемент під індексом 2 3
Останній елемент 3


Зрізи: 

In [14]:
arr = np.array([1, 2, 3, 4, 5, 6])
print("Від 0 індекса включно по 2 невключно", arr[0:2])
print("до 2 індекса невключно", arr[:2])
print("з 2 індекса невключно", arr[2:])

Від 0 індекса включно по 2 невключно [1 2]
до 2 індекса невключно [1 2]
з 2 індекса невключно [3 4 5 6]


Індексація у двовимірних масивах: 

In [15]:
arr = np.array([[1, 2, 3], [4, 5, 6]])
print("перший рядок, другий стовпець", arr[0, 1])

перший рядок, другий стовпець 2


### Форми та Перетворення Масивів

Форма масиву: 

In [16]:
arr = np.array([[1, 2, 3], [4, 5, 6]])
print(arr.shape)

(2, 3)

Зміна форми: 

In [17]:
arr = np.array([[1, 2, 3], [4, 5, 6]])
print(arr.reshape(3, 2))

[[1 2]
 [3 4]
 [5 6]]


### Типи Даних:

Дізнатися тип даних масиву: 

In [21]:
arr = np.array([[1, 2, 3], [4, 5, 6]])
arr.dtype

dtype('int64')

Задавання типу даних при створенні: 

In [22]:
arr = np.array([[1, 2, 3], [4, 5, 6]], dtype=np.float64)
arr.dtype

dtype('float64')

### Арифметичні Операції:

Операції з числами: 
    

In [25]:
arr = np.array([[1, 2, 3], [4, 5, 6]])
arr + 1

array([[2, 3, 4],
       [5, 6, 7]])

In [26]:
arr = np.array([[1, 2, 3], [4, 5, 6]])
arr - 1

array([[0, 1, 2],
       [3, 4, 5]])

In [27]:
arr = np.array([[1, 2, 3], [4, 5, 6]])
arr * 2

array([[ 2,  4,  6],
       [ 8, 10, 12]])

In [28]:
arr = np.array([[1, 2, 3], [4, 5, 6]])
arr / 2.

array([[0.5, 1. , 1.5],
       [2. , 2.5, 3. ]])

Операції з масивами: 
    

In [30]:
arr1 = np.array([[1, 2, 3], [4, 5, 6]])
arr2 = np.array([[2, 3, 4], [1, 3, 2]])
arr1 + arr2

array([[3, 5, 7],
       [5, 8, 8]])

In [32]:
arr1 = np.array([[1, 2, 3], [4, 5, 6]])
arr2 = np.array([[6, 1, 3], [1, 3, 2]])
arr1 - arr2

array([[-5,  1,  0],
       [ 3,  2,  4]])

### Логічні Операції:
Порівняння: 

In [33]:
arr = np.array([[1, 2, 3], [4, 5, 6]])
arr > 2

array([[False, False,  True],
       [ True,  True,  True]])

Фільтрація елементів по умові

In [40]:
arr = np.array([[1, 2, 3], [4, 5, 6]])
arr[arr > 2]

array([3, 4, 5, 6])

In [34]:
arr = np.array([[1, 2, 3], [4, 5, 6]])
arr == 5

array([[False, False, False],
       [False,  True, False]])

Логічні операції: 

In [35]:
arr = np.array([[1, 2, 3], [4, 5, 6]])
np.logical_and(arr > 2, arr < 5)

array([[False, False,  True],
       [ True, False, False]])

In [37]:
arr = np.array([[1, 2, 3], [4, 5, 6]])
np.logical_or(arr > 1, arr < 1)

array([[False,  True,  True],
       [ True,  True,  True]])

### Завдання для Самоперевірки

1. Створіть одновимірний масив з елементами від 5 до 15. Змініть форму цього масиву на 2x5. Отримайте всі елементи, які більші за 10. Порахуйте кількість елементів у вашому масиві, які діляться на 3 без залишку.


2. Створіть масив 3x3 з нулів, а потім замініть центральний елемент на 1. Відніміть 2 від усіх елементів.
3. Створіть два масиви різних форм і спробуйте виконати арифметичні операції над ними. Поясніть результат.

# Продвинуті Операції з Масивами

### Агрегація Даних:

Сума: 

In [3]:
arr = np.array([[1, 2, 3], [4, 5, 6]])
np.sum(arr)

21

Мінімум та максимум: 

In [4]:
arr = np.array([[1, 2, 3], [4, 5, 6]])
np.min(arr)

1

In [5]:
arr = np.array([[1, 2, 3], [4, 5, 6]])
np.max(arr)

6

Середнє та медіана: 

In [6]:
arr = np.array([[1, 2, 3], [4, 5, 6]])
np.mean(arr)

3.5

In [9]:
arr = np.array([[1, 2, 3], [3, 4, 6]])
np.median(arr)

3.0

Стандартне відхилення: 

In [10]:
arr = np.array([[1, 2, 3], [4, 5, 6]])
np.std(arr)

1.707825127659933

### Трансформації Масивів:

Зміна форми (не змінюючи дані): 

In [11]:
arr = np.array([[1, 2, 3], [4, 5, 6]])
arr.reshape((3, 2))

array([[1, 2],
       [3, 4],
       [5, 6]])

Вирівнювання масиву в один рядок: 

In [12]:
arr = np.array([[1, 2, 3], [4, 5, 6]])
arr.ravel()

array([1, 2, 3, 4, 5, 6])

In [13]:
arr = np.array([[1, 2, 3], [4, 5, 6]])
arr.flatten()

array([1, 2, 3, 4, 5, 6])

### Об'єднання та Розділення Масивів:

Об'єднання: 

In [14]:
arr1 = np.array([[1, 2, 3], [4, 5, 6]])
arr2 = np.array([[7, 8, 9], [10, 11, 12]])
np.concatenate([arr1, arr2])

array([[ 1,  2,  3],
       [ 4,  5,  6],
       [ 7,  8,  9],
       [10, 11, 12]])

In [15]:
arr1 = np.array([[1, 2, 3], [4, 5, 6]])
arr2 = np.array([[7, 8, 9], [10, 11, 12]])
np.vstack([arr1, arr2])

array([[ 1,  2,  3],
       [ 4,  5,  6],
       [ 7,  8,  9],
       [10, 11, 12]])

In [17]:
arr1 = np.array([[1, 2, 3], [7, 8, 9]])
arr2 = np.array([[4, 5, 6], [10, 11, 12]])
np.hstack([arr1, arr2])

array([[ 1,  2,  3,  4,  5,  6],
       [ 7,  8,  9, 10, 11, 12]])

Розділення: 

In [21]:
arr = np.array([1, 2, 3, 4, 5, 6])
np.split(arr, [1, 3])

[array([1]), array([2, 3]), array([4, 5, 6])]

In [25]:
arr = np.array([[1, 2, 3], [4, 5, 6]])
np.hsplit(arr, [1])

[array([[1],
        [4]]),
 array([[2, 3],
        [5, 6]])]

In [27]:
arr = np.array([[1, 2, 3], [4, 5, 6], [7, 8, 9]])
np.vsplit(arr, [1])

[array([[1, 2, 3]]),
 array([[4, 5, 6],
        [7, 8, 9]])]

### Індексація та Слайсинг Заумовлених Масивів:

Булева індексація: 

In [28]:
arr = np.array([[1, 2, 3], [4, 5, 6]])
arr[arr > 5]

array([6])

Функції np.where: 

In [29]:
arr = np.array([[1, 2, 3], [4, 5, 6]])
np.where(arr > 5, 1, 0)

array([[0, 0, 0],
       [0, 0, 1]])

### Векторизація Функцій:

Застосування функцій до кожного елемента: 

In [34]:
arr = np.array([[1, 2, 3], [4, 5, 6]])
arr**2 + arr

array([[ 2,  6, 12],
       [20, 30, 42]])

In [35]:
def func(x):
    return x**2 + x

np.vectorize(func)(arr)

array([[ 2,  6, 12],
       [20, 30, 42]])

### Бінарні Операції та Порівняння:

Порівняння масивів: 

In [37]:
arr1 = np.array([[1, 2, 3], [7, 8, 9]])
arr2 = np.array([[1, 5, 6], [10, 8, 12]])
np.equal(arr1, arr2)

array([[ True, False, False],
       [False,  True, False]])

In [39]:
arr1 = np.array([[1, 9, 3], [7, 8, 13]])
arr2 = np.array([[4, 5, 6], [10, 11, 12]])
np.greater(arr1, arr2)

array([[False,  True, False],
       [False, False,  True]])

### Завдання для Самоперевірки
* Створіть два масиви з випадковими числами та знайдіть їхню суму, мінімум та максимум.
* Змініть форму одного з масивів і об'єднайте їх вертикально.
* Розділіть отриманий в попередньому пункті масив на три рівні частини.
* Створіть масив і знайдіть всі елементи, які більші за середнє значення цього масиву.
* Створіть власну функцію, яка змінює значення елемента, і застосуйте її до масиву з використанням np.vectorize.
* Порівняйте два масиви і знайдіть елементи, які є в обох масивах.
* Створіть масив і реалізуйте умовне присвоєння значень за допомогою np.where.

### Операції Лінійної Алгебри в NumPy


#### Множення Матриць:
- **Використання**: `np.dot(a, b)` або `a @ b` для множення матриць `a` і `b`.


Множення матриць - це основна операція в лінійній алгебрі, яка використовується для комбінування даних з двох матриць, щоб отримати нову матрицю. 

Example 1:

![assert/animated-gif-determinant-of-2x2-matrix.gif](assert/animated-gif-determinant-of-2x2-matrix.gif)


Example 2:

![assert/multiply_matrices.gif](assert/multiply_matrices.gif)

У цьому прикладі ми множимо дві квадратні матриці a і b, кожна з розміром 2x2. Результатом буде нова матриця, де кожен елемент обчислюється як скалярний добуток відповідних рядків з матриці a і стовпців з матриці b.



In [43]:
import numpy as np

# Створення двох квадратних матриць розміром 2x2
a = np.array([[1, 2],
              [3, 4]])

b = np.array([[5, 6],
              [7, 8]])

# Множення матриць a і b
result = np.dot(a, b)

print(result)

[[19 22]
 [43 50]]


У цьому випадку ми множимо матрицю matrix розміром 2x3 на вектор vector розміром 3. Результатом буде новий вектор, де кожен елемент обчислюється як скалярний добуток відповідного рядка з матриці і вектора.

In [42]:
import numpy as np

# Створення матриці розміром 2x3 і вектора розміром 3
matrix = np.array([[1, 2, 3],
                   [4, 5, 6]])

vector = np.array([7, 8, 9])

# Множення матриці на вектор
result = np.dot(matrix, vector)

print(result)

[ 50 122]


#### Визначник та Ранг Матриці:
- **Визначник**: `np.linalg.det(a)`.
- **Ранг матриці**: `np.linalg.matrix_rank(a)`.


Визначник і ранг матриці - це важливі характеристики матриць, які використовуються в лінійній алгебрі та чисельних методах.

<b>Визначник матриці</b> - це числова характеристика матриці, яка обчислюється для квадратної матриці. В даному прикладі, визначник матриці a дорівнює 10. Визначник вказує на те, як матриця впливає на масштаб області векторів у просторі, і він є корисним, наприклад, при розв'язанні систем лінійних рівнянь і обчисленні оберненої матриці.



2 dimensional:

![assert/animated-gif-determinant-of-2x2-matrix.gif](assert/animated-gif-determinant-of-2x2-matrix.gif)

3 dimensional

![assert/animated-formula-determinant-3x3-matrix.gif](assert/animated-formula-determinant-3x3-matrix.gif)

4 dimensional

![assert/suud1.1.gif](assert/suud1.1.gif)

In [44]:
import numpy as np

# Створення квадратної матриці 2x2
a = np.array([[3, 1],
              [2, 4]])

# Обчислення визначника матриці a
determinant = np.linalg.det(a)

print("Визначник матриці a:", determinant)

Визначник матриці a: 10.000000000000002


<b>Ранг</b> матриці вказує на максимальну кількість лінійно незалежних рядків (або стовпців) у матриці. У цьому прикладі, ранг матриці b дорівнює 2, що означає, що є два лінійно незалежних рядки у цій матриці. Ранг матриці є важливою характеристикою при розв'язанні систем лінійних рівнянь і визначенні розмірності простору, що охоплюється матрицею.

2 dimensional


![assert/1_XXmAVU0DypKnwGrg5nQE1Q.webp](assert/1_XXmAVU0DypKnwGrg5nQE1Q.webp)


4 dimensional

![assert/1__VKG-MvESmtQSmwc7tkIkA.webp](assert/1__VKG-MvESmtQSmwc7tkIkA.webp)

In [45]:
import numpy as np

# Створення матриці розміром 3x3
b = np.array([[1, 2, 3],
              [4, 5, 6],
              [7, 8, 9]])

# Обчислення рангу матриці b
rank = np.linalg.matrix_rank(b)

print("Ранг матриці b:", rank)

Ранг матриці b: 2


#### Слід Матриці:
- **Слід (сума елементів на головній діагоналі)**: `np.trace(a)`.


Слід матриці - це сума всіх елементів, які розташовані на головній діагоналі квадратної матриці. Визначення сліду матриці виглядає так:

Слід(A) = a[0,0] + a[1,1] + ... + a[n-1,n-1], де A - квадратна матриця розміром n x n.

Давайте розглянемо приклад та пояснення обчислення сліду матриці за допомогою функції np.trace(a) в бібліотеці NumPy.

In [49]:
import numpy as np

# Створення квадратної матриці 3x3
a = np.array([[1, 2, 3],
              [4, 5, 6],
              [7, 8, 9]])

# Обчислення сліду матриці a
trace = np.trace(a)

print("Слід матриці a:", trace)

Слід матриці a: 15


У цьому прикладі ми маємо квадратну матрицю a розміром 3x3. Слід матриці обчислюється як сума елементів на головній діагоналі, тобто 1 + 5 + 9, що дорівнює 15. Отже, слід матриці a дорівнює 15.

Слід матриці має декілька важливих використань в лінійній алгебрі та матричних операціях. Наприклад, він може бути корисним при обчисленні власних значень та власних векторів матриці, а також у визначенні детермінанта матриці.

#### Рішення Лінійних Систем Рівнянь:
- **Формула**: `np.linalg.solve(A, b)`, де `A` - матриця коефіцієнтів, а `b` - вектор вільних членів.


Рішення лінійних систем рівнянь є важливою задачею в лінійній алгебрі та чисельних методах. Для цього використовується формула np.linalg.solve(A, b), де A - матриця коефіцієнтів системи рівнянь, а b - вектор вільних членів. Давайте розглянемо приклад та пояснення цієї операції.


Приклад:
Розглянемо наступну систему лінійних рівнянь:
```
2x + y = 5
3x - 2y = 8
```
Ми можемо представити цю систему у вигляді матриці коефіцієнтів A і вектора вільних членів b:

In [50]:
import numpy as np

# Матриця коефіцієнтів A
A = np.array([[2, 1],
              [3, -2]])

# Вектор вільних членів b
b = np.array([5, 8])

# Рішення системи рівнянь
solution = np.linalg.solve(A, b)

print("Рішення системи рівнянь (x, y):", solution)

Рішення системи рівнянь (x, y): [ 2.57142857 -0.14285714]


У цьому прикладі ми використали функцію np.linalg.solve(A, b) для знаходження розв'язку системи лінійних рівнянь. Результатом є вектор, який містить значення змінних x та y, що задовольняють обидві рівняння одночасно. У нашому випадку розв'язок буде x = 2 і y = 1.

Функція np.linalg.solve розв'язує системи лінійних рівнянь ефективним чисельним методом і дуже корисна в багатьох областях, включаючи науку про дані, інженерію та фізику.

#### Власні Значення та Власні Вектори:
- **Функція**: `np.linalg.eig(a)` повертає кортеж з власних значень та власних векторів матриці `a`.


Власні значення і власні вектори є важливими концепціями в лінійній алгебрі та матричних обчисленнях. Вони використовуються для аналізу та розв'язання численних задач у багатьох областях науки та інженерії. Функція np.linalg.eig(a) в бібліотеці NumPy дозволяє знайти власні значення і власні вектори матриці a. Давайте розглянемо приклади і пояснення цих концепцій.

Приклад 1: Знаходження власних значень і власних векторів квадратної матриці

In [52]:
import numpy as np

# Створення квадратної матриці 2x2
a = np.array([[3, 1],
              [2, 2]])

# Знаходження власних значень і власних векторів
eigenvalues, eigenvectors = np.linalg.eig(a)

print("Власні значення:", eigenvalues)
print("Власні вектори:")
print(eigenvectors)

Власні значення: [4. 1.]
Власні вектори:
[[ 0.70710678 -0.4472136 ]
 [ 0.70710678  0.89442719]]


Власні значення (eigenvalues) матриці a - це числа, які характеризують, як матриця перетворює простір. Власні вектори (eigenvectors) - це вектори, які залишають свій напрям після цього перетворення.

Приклад 2: Знаходження власних значень і власних векторів для симетричної матриці

In [53]:
import numpy as np

# Створення симетричної матриці 3x3
b = np.array([[1, 2, 2],
              [2, 3, 4],
              [2, 4, 5]])

# Знаходження власних значень і власних векторів
eigenvalues, eigenvectors = np.linalg.eig(b)

print("Власні значення:", eigenvalues)
print("Власні вектори:")
print(eigenvectors)

Власні значення: [ 9.09783468  0.28620826 -0.38404294]
Власні вектори:
[[-0.32798528 -0.73697623  0.59100905]
 [-0.59100905 -0.32798528 -0.73697623]
 [-0.73697623  0.59100905  0.32798528]]


У цьому прикладі ми розглядаємо симетричну матрицю b. В такому випадку власні значення завжди є дійсними числами, і власні вектори є ортогональними один одному.

Знання власних значень і власних векторів може бути корисним у багатьох задачах, таких як аналіз коливань в фізиці, розпізнавання образів в обробці зображень, розв'язання систем лінійних диференціальних рівнянь та багатьох інших.

#### Обернена Матриця:
- **Функція**: `np.linalg.inv(a)`.


Обернена матриця - це матриця, яка, помножена на вихідну матрицю, дає одиничну матрицю. Обернена матриця дозволяє розв'язувати системи лінійних рівнянь, знаходити розв'язки та виконувати інші математичні операції. У бібліотеці NumPy функція np.linalg.inv(a) використовується для знаходження оберненої матриці. Давайте розглянемо приклади та пояснення цієї операції.

In [54]:
import numpy as np

# Створення квадратної матриці 2x2
a = np.array([[3, 1],
              [2, 4]])

# Знаходження оберненої матриці
inverse_a = np.linalg.inv(a)

print("Вихідна матриця:")
print(a)
print("Обернена матриця:")
print(inverse_a)

Вихідна матриця:
[[3 1]
 [2 4]]
Обернена матриця:
[[ 0.4 -0.1]
 [-0.2  0.3]]


Обернена матриця для матриці a обчислюється за допомогою функції np.linalg.inv(a). В результаті ми отримуємо матрицю, яка, помножена на вихідну матрицю a, дорівнює одиничній матриці. Обернена матриця дозволяє розв'язувати системи лінійних рівнянь та інші математичні операції, але важливо враховувати, що не всі матриці мають обернену.

Приклад 2: Застосування оберненої матриці для розв'язання системи рівнянь

In [55]:
import numpy as np

# Створення матриці коефіцієнтів A та вектора вільних членів b
A = np.array([[3, 1],
              [2, 4]])

b = np.array([9, 10])

# Знаходження оберненої матриці A
inverse_A = np.linalg.inv(A)

# Розв'язання системи рівнянь Ax = b
x = np.dot(inverse_A, b)

print("Розв'язок системи рівнянь (x1, x2):", x)

Розв'язок системи рівнянь (x1, x2): [2.6 1.2]


У цьому прикладі ми спочатку знаходимо обернену матрицю inverse_A, а потім використовуємо її для розв'язання системи лінійних рівнянь Ax = b. Розв'язок цієї системи дає нам значення змінних x1 і x2.

#### Завдання для Самоперевірки
1. **Створіть** дві матриці та **виконайте** їх множення.
2. **Розрахуйте** визначник та ранг згенерованої матриці.
3. **Знайдіть** слід матриці.
4. **Розв'яжіть** лінійну систему рівнянь, створивши власну систему з двох рівнянь і двох невідомих.
5. **Знайдіть** власні значення та власні вектори для квадратної матриці.
6. **Створіть** квадратну матрицю та **знайдіть** її обернену матрицю.
7. **Перевірте**, чи дійсно обернена матриця, отримана у попередньому пункті, є правильною (перемножте її з оригінальною матрицею).


### Обробка Даних в NumPy


#### Використання Масивів для Обробки Даних:
- **Завантаження даних у масиви**: Використання `np.loadtxt()` або `np.genfromtxt()` для читання даних з файлів.
- **Застосування математичних та статистичних функцій**: `np.mean()`, `np.median()`, `np.std()`, `np.percentile()`.


Використання масивів для обробки даних є дуже поширеною задачею у наукових областях, програмуванні та аналізі даних. NumPy, бібліотека для чисельних обчислень в Python, надає засоби для завантаження, обробки і аналізу даних в масивах. Давайте розглянемо приклади та пояснення використання масивів для цих завдань.

###### Збереження данних 

In [59]:
x = np.arange(0.0,5.0,1.0)
np.savetxt('data.txt', x)

###### Завантаження даних у масиви:

Приклад 1: Завантаження даних з текстового файлу за допомогою np.loadtxt().

In [60]:
import numpy as np

# Завантаження даних з файлу "data.txt" (розділені пробілами) у масив
data = np.loadtxt("data.txt")

print("Завантажені дані:")
print(data)

Завантажені дані:
[0. 1. 2. 3. 4.]


У цьому прикладі np.loadtxt() використовується для завантаження даних з файлу "data.txt" у масив data. Функція автоматично визначає типи даних та роздільники між значеннями.


###### Застосування математичних та статистичних функцій:

Приклад 2: Застосування функцій np.mean(), np.median(), np.std(), np.percentile() до масиву даних.

In [61]:
import numpy as np

# Створення масиву даних
data = np.array([12, 15, 18, 22, 30, 35, 40, 55, 60, 75])

# Обчислення середнього значення
mean_value = np.mean(data)

# Обчислення медіани
median_value = np.median(data)

# Обчислення стандартного відхилення
std_deviation = np.std(data)

# Обчислення 25-го та 75-го перцентилів
percentile_25 = np.percentile(data, 25)
percentile_75 = np.percentile(data, 75)

print("Середнє значення:", mean_value)
print("Медіана:", median_value)
print("Стандартне відхилення:", std_deviation)
print("25-й перцентиль:", percentile_25)
print("75-й перцентиль:", percentile_75)

Середнє значення: 36.2
Медіана: 32.5
Стандартне відхилення: 20.118648065911387
25-й перцентиль: 19.0
75-й перцентиль: 51.25


У цьому прикладі ми використовуємо функції NumPy для обчислення різних статистичних параметрів масиву data, таких як середнє значення, медіана, стандартне відхилення та перцентилі.

Ці інструменти дозволяють легко і ефективно обробляти та аналізувати дані в Python, що робить їх корисними для наукових досліджень, аналізу даних та інших завдань.


* Середнє значення: 36.2
Уяви, що ти ділиш цукерки між друзями. Середнє значення - це коли ти збираєш усі цукерки разом і ділиш їх порівну між усіма. У нашому випадку, якби цукерки були числами, кожен отримав би приблизно 36 цукерок.
* Медіана: 32.5
Уяви, що ти з друзями стоїш у ряд за зростом. Медіана - це зріст того, хто стоїть посередині. Тут це число 32.5, тобто половина чисел менша за 32.5, а половина - більша.
* Стандартне відхилення: 20.12
Це як "середня відстань" усіх чисел від середнього значення. Якщо уявити, що всі числа - це діти на ігровому майданчику, то стандартне відхилення показує, наскільки далеко вони розбіглися від центру майданчика.
* 25-й перцентиль: 19.0
Уяви, що ти роздаєш морозиво 100 дітям. 25-й перцентиль - це кількість морозива у 25-ї дитини, якщо рахувати від того, хто отримав найменше. Тут це 19, тобто 25 дітей отримали 19 або менше морозива.
* 75-й перцентиль: 51.25
А це кількість морозива у 75-ї дитини. 51.25 означає, що 75 дітей отримали 51.25 або менше морозива, а 25 дітей - більше.

#### Функції для Статистичного Аналізу:
- **Розрахунок кореляції**: `np.corrcoef()`.
- **Розрахунок коваріації**: `np.cov()`.
- **Стандартизація даних**: `(arr - np.mean(arr)) / np.std(arr)`.


Статистичний аналіз даних важливий для виявлення зв'язків і закономірностей у наборі даних. Функції NumPy дозволяють виконувати різні операції для статистичного аналізу, такі як розрахунок кореляції, коваріації та стандартизації даних. Давайте розглянемо приклади та пояснення використання цих функцій.


* Кореляція
Уяви, що ти з другом їсте морозиво. Кореляція показує, чи ви їсте його однаково. Якщо ти їси багато, і твій друг теж їсть багато, або ти їси мало, і друг теж мало - це позитивна кореляція. Якщо ти їси багато, а друг мало (або навпаки) - це негативна кореляція. Якщо ваші звички не пов'язані - кореляції немає.
* Коваріація
Це схоже на кореляцію, але трохи складніше. Уяви, що ви з другом стрибаєте на батуті. Коваріація показує, чи стрибаєте ви в одному напрямку. Якщо ви обидва стрибаєте високо або обидва низько - коваріація позитивна. Якщо один високо, а інший низько - негативна.
* Стандартизація даних
Уяви, що в класі є високі і низькі діти. Стандартизація - це ніби всіх дітей "вирівняли" за зростом. Після цього ми можемо легко порівняти, хто вищий чи нижчий від середнього зросту в класі.

###### Розрахунок кореляції:

Приклад 1: Розрахунок кореляційної матриці за допомогою np.corrcoef().

In [62]:
import numpy as np

# Створення двох масивів з оцінками для двох ознак (наприклад, оцінки за математику і фізику)
math_scores = np.array([90, 85, 88, 92, 78])
physics_scores = np.array([88, 87, 84, 90, 82])

# Розрахунок кореляційної матриці
correlation_matrix = np.corrcoef(math_scores, physics_scores)

print("Кореляційна матриця:")
print(correlation_matrix)

Кореляційна матриця:
[[1.         0.83742252]
 [0.83742252 1.        ]]


У цьому прикладі ми використовуємо функцію np.corrcoef() для розрахунку кореляційної матриці між двома масивами даних, які представляють оцінки за математику та фізику. Кореляційна матриця показує кореляційні коефіцієнти між ознаками. Значення на головній діагоналі будуть 1, оскільки кореляція між ознаками та самих собою завжди дорівнює 1.

###### Розрахунок коваріації:

Приклад 2: Розрахунок коваріації між двома масивами за допомогою np.cov().

In [63]:
import numpy as np

# Створення двох масивів з оцінками для двох ознак (наприклад, оцінки за математику і фізику)
math_scores = np.array([90, 85, 88, 92, 78])
physics_scores = np.array([88, 87, 84, 90, 82])

# Розрахунок коваріації між оцінками за математику і фізику
covariance = np.cov(math_scores, physics_scores)

print("Коваріація між математикою і фізикою:")
print(covariance)

Коваріація між математикою і фізикою:
[[29.8 14.6]
 [14.6 10.2]]


У цьому прикладі ми використовуємо функцію np.cov() для розрахунку коваріації між двома масивами даних, які представляють оцінки за математику та фізику. Коваріація вказує на ступінь лінійного відношення між двома ознаками.

###### Стандартизація даних:

Приклад 3: Стандартизація масиву даних за допомогою формули (arr - np.mean(arr)) / np.std(arr).

In [65]:
import numpy as np

# Створення масиву з випадковими даними
data = np.array([45, 60, 72, 55, 48, 68, 80, 62, 58, 70])

# Стандартизація даних
standardized_data = (data - np.mean(data)) / np.std(data)

print("Стандартизовані дані:")
print(standardized_data)

Стандартизовані дані:
[-1.61837995 -0.17339785  0.98258783 -0.65505855 -1.32938353  0.59725927
  1.75324495  0.01926643 -0.36606213  0.78992355]


У цьому прикладі ми застосовуємо формулу стандартизації до масиву даних. Спершу віднімаємо середнє значення масиву (np.mean(data)) від кожного значення, а потім ділимо отримане значення на стандартне відхилення масиву (np.std(data)). Результат - стандартизовані дані, які мають середнє значення 0 і стандартне відхилення 1.

#### Обробка Пропущених Значень:
- **Виявлення пропущених значень**: `np.isnan(arr)`.
- **Заповнення пропущених значень**: `arr[np.isnan(arr)] = value`.


Обробка пропущених значень є важливим етапом в аналізі та обробці даних. NumPy надає зручні засоби для виявлення та заповнення пропущених значень в масивах даних. Давайте розглянемо приклади та пояснення використання цих функцій.

###### Виявлення пропущених значень:

Приклад 1: Виявлення пропущених значень в масиві за допомогою np.isnan().

In [66]:
import numpy as np

# Створення масиву з даними, включаючи пропущене значення (NaN)
data = np.array([12, 18, 15, np.nan, 24, np.nan, 30, 27])

# Виявлення пропущених значень
missing_values = np.isnan(data)

print("Масив пропущених значень:")
print(missing_values)

Масив пропущених значень:
[False False False  True False  True False False]


У цьому прикладі функція np.isnan() використовується для виявлення пропущених значень в масиві data. Функція повертає масив булевих значень, де True вказує на наявність пропущених значень, а False - на наявність дійсних числових значень.

###### Заповнення пропущених значень:

Приклад 2: Заповнення пропущених значень певним значенням за допомогою індексації.

In [67]:
import numpy as np

# Створення масиву з даними, включаючи пропущене значення (NaN)
data = np.array([12, 18, 15, np.nan, 24, np.nan, 30, 27])

# Заповнення пропущених значень значенням 0
data[np.isnan(data)] = 0

print("Масив з заповненими пропущеними значеннями:")
print(data)

Масив з заповненими пропущеними значеннями:
[12. 18. 15.  0. 24.  0. 30. 27.]


У цьому прикладі ми використовуємо індексацію масиву data[np.isnan(data)] для вибору всіх пропущених значень (де np.isnan(data) є True) і присвоюємо їм значення 0. Як результат, ми отримуємо масив, в якому пропущені значення заповнені заданим значенням (0 в цьому випадку).

Ці функції дозволяють легко виявляти та обробляти пропущені значення в масивах даних, що є важливим завданням в аналізі та обробці даних.

#### Сортування та Фільтрація:
- **Сортування масивів**: `np.sort(arr)`.
- **Умовна фільтрація**: Використання булевих масивів для фільтрації даних.


Сортування та фільтрація є важливими операціями при обробці та аналізі даних. NumPy надає зручні засоби для сортування масивів та фільтрації даних з використанням булевих масивів. Давайте розглянемо приклади та пояснення цих операцій.

##### Сортування масивів:

Приклад 1: Сортування масиву за допомогою np.sort().

In [69]:
import numpy as np

# Створення масиву з випадковими числами
arr = np.array([35, 12, 55, 7, 24, 48, 18, 62])

# Сортування масиву в порядку зростання
sorted_arr = np.sort(arr)

print("Відсортований масив:")
print(sorted_arr)

Відсортований масив:
[ 7 12 18 24 35 48 55 62]


У цьому прикладі ми використовуємо функцію np.sort() для сортування масиву arr в порядку зростання. Функція повертає новий масив, в якому значення впорядковані.

###### Умовна фільтрація:

Приклад 2: Фільтрація масиву на основі умови за допомогою булевого масиву.

In [70]:
import numpy as np

# Створення масиву з випадковими числами
arr = np.array([35, 12, 55, 7, 24, 48, 18, 62])

# Створення булевого масиву, де True відповідає числам більше 30
condition = arr > 30

# Фільтрація масиву на основі умови
filtered_arr = arr[condition]

print("Масив, що відповідає умові:")
print(filtered_arr)

Масив, що відповідає умові:
[35 55 48 62]


У цьому прикладі ми створюємо булевий масив condition, де True відповідає елементам масиву arr, які більше 30. Потім ми використовуємо цей булевий масив для фільтрації arr за допомогою індексації. Результат - масив, що містить тільки ті значення, які відповідають умові.

Ці операції дозволяють ефективно сортувати та фільтрувати дані, що є важливими завданнями в аналізі та обробці даних.

#### Векторизація:
- **Використання векторизованих операцій** для ефективної обробки даних без необхідності використання циклів.


Векторизація - це техніка, яка дозволяє виконувати операції на цілих масивах даних одночасно, без необхідності використання циклів або інших ітераційних конструкцій. В NumPy, бібліотеці для чисельних обчислень в Python, векторизовані операції реалізуються за допомогою масивів, що дозволяє вам ефективно обробляти великі об'єми даних. Давайте розглянемо приклади та пояснення векторизації.

Приклад 1: Додавання двох масивів без векторизації.

In [71]:
import numpy as np

# Створення двох масивів
a = np.array([1, 2, 3, 4, 5])
b = np.array([10, 20, 30, 40, 50])

# Додавання елементів масивів один до одного
result = np.empty_like(a)
for i in range(len(a)):
    result[i] = a[i] + b[i]

print("Результат додавання без векторизації:")
print(result)

Результат додавання без векторизації:
[11 22 33 44 55]


У цьому прикладі ми додаємо відповідні елементи двох масивів a і b за допомогою циклу. Це працює, але для великих масивів це може бути повільно.

Приклад 2: Додавання двох масивів з векторизацією.

In [72]:
import numpy as np

# Створення двох масивів
a = np.array([1, 2, 3, 4, 5])
b = np.array([10, 20, 30, 40, 50])

# Додавання елементів масивів за допомогою векторизованої операції
result = a + b

print("Результат додавання з векторизацією:")
print(result)

Результат додавання з векторизацією:
[11 22 33 44 55]


У цьому прикладі ми просто додаємо масиви a і b, і операція виконується одночасно для всіх елементів. Це значно підвищує швидкість обчислень, особливо при роботі з великими об'ємами даних.

Векторизація дозволяє писати більш зрозумливий та ефективний код, зменшує кількість потрібних ітерацій та полегшує обробку даних. У NumPy і інших бібліотеках для наукових обчислень векторизація є невід'ємною частиною для оптимізації обчислень.

#### Завдання для Самоперевірки
1. **Завантажте датасет** (наприклад, з файлу CSV) у NumPy масив і **виведіть основні статистики** (середнє, медіана, стандартне відхилення).
2. **Розрахуйте кореляцію** між двома стовпцями датасету.
3. **Стандартизуйте один із стовпців** датасету.
4. **Ідентифікуйте та заповніть пропущені значення** в одному із стовпців.
5. **Виконайте сортування** датасету за значеннями одного зі стовпців.
6. **Фільтруйте дані**, використовуючи булеві умови (наприклад, виберіть всі записи, де значення в стовпці більше певного числа).
7. **Застосуйте власну функцію** до елементів масиву з допомогою `np.vectorize`.



# Додаткові Матеріали та Ресурси для Вивчення NumPy

## Офіційна Документація NumPy
- **NumPy Documentation**: Офіційна документація є першим та найважливішим ресурсом. Вона включає пояснення всіх функцій, класів, і модулів, а також численні приклади.

## Онлайн Курси та Навчальні Посібники
- **Coursera, Udemy, edX**: Багато курсів по програмуванню в Python включають розділи про NumPy.
- **Khan Academy**: Курси з математики та статистики, які можна використовувати для кращого розуміння фонових знань.

## Книги
- **"Python for Data Analysis"** автора Wes McKinney: Книга добре пояснює використання NumPy в контексті аналізу даних.
- **"Numerical Python"** автора Robert Johansson: Включає більш глибоке розуміння чисельних методів у NumPy.

## Інтерактивні Платформи
- **DataCamp, Codecademy**: Інтерактивні платформи, де можна практикувати роботу з NumPy.

## Відеоуроки
- **YouTube**: Численні підручники та лекції, які можна знайти, використовуючи пошук "NumPy tutorials".

## Форуми та Спільноти
- **Stack Overflow**: Велика спільнота, де можна ставити питання та отримувати відповіді.
- **GitHub**: Джерело для вивчення коду та участі у проектах, які використовують NumPy.
